### PR TITLE
Handle the case of large step causing single pt extents

### PR DESCRIPTION
**Pull request by gouthamve**
We noticed that we were ignoring samples if the step is 24h and this
fixes it.

### CODE REVIEW

**Comment by bboreham** [comment]
I see that this improves the symptom, however it now does something undesirable: it makes two queries for parts of the range before and after the cached extent, with no gap between them since the extent is of zero width.
Can you special-case this so the cached value is added but only one query is made?

**Comment by gouthamve** [comment]
@bboreham Handled the special case differently, WDYT?

**Comment by bboreham** [comment]
Still looks like it will do the same 2 queries.  Try changing the test so the cached value is at 101.

**Comment by bboreham** [comment]
Wait how does this even work?  Your test is for a zero-length range (`Start: 100, End: 100`); does that happen in practice?

**Comment by gouthamve** [comment]
It happens when the step is more than 50% of the `interval`. Or basically large steps. In the case in our production it was `24hr` step.

https://play.golang.org/p/rc0pex0NISJ

https://github.com/cortexproject/cortex/blob/master/pkg/querier/queryrange/split_by_interval.go#L63-L86

**Comment by bboreham** [comment]
OK can you have different handling for `req.GetStart() == req.GetEnd()` ? 

AFAICS for a longer request your change will put back the conditions I was fixing in #3653

**Comment by gouthamve** [comment]
> AFAICS for a longer request your change will put back the conditions I was fixing in #3653

Basically when you have `req.GetStart() == req.GetEnd()`, it is impossible to have the conditions of #3653. When start and end are the same, you basically have a _single_ step in that range. If there is a hit, we pick that extent (no other extent can exist). We basically can never have many small requests.

**Comment by bboreham** [comment]
OK but this PR changes it for all requests.

**Comment by bboreham** [comment]
Ah, no it doesn't!

Sorry, I did not understand the comment at all, and didn't spot how the test had changed.

Still think it would be clearer to special-case `req.GetStart() == req.GetEnd()` outside of the loop.

File: pkg/querier/queryrange/results_cache.go [hunk 1]
```diff
@@ -479,8 +479,13 @@ func (s resultsCache) partition(req Request, extents []Extent) ([]Request, []Res
 		if extent.GetEnd() < start || extent.Start > req.GetEnd() {
 			continue
 		}
-		// If this extent is tiny, discard it: more efficient to do a few larger queries
-		if extent.End-extent.Start < s.minCacheExtent {
+
+		// If this extent is tiny, discard it: more efficient to do a few larger queries.
+
+		// However if the step is large enough, the split_query_by_interval middlware would generate a query with same start and end.
```
Comment (pracucci) [inline pkg/querier/queryrange/results_cache.go:485]:
```suggestion
		// However if the step is large enough, the split_query_by_interval middleware would generate a query with same start and end.
```

File: pkg/querier/queryrange/results_cache.go [hunk 2]
```diff
@@ -504,6 +509,12 @@ func (s resultsCache) partition(req Request, extents []Extent) ([]Request, []Res
 		requests = append(requests, r)
 	}
 
+	// If start and end are the same (valid in promql), start == req.GetEnd() and we won't do the query.
+	// But we should only do the request if we don't have a valid cached response for it.
+	if req.GetStart() == req.GetEnd() && len(cachedResponses) == 0 {
```
Comment (pracucci) [inline pkg/querier/queryrange/results_cache.go:514]: Shouldn't we just change the previous `if start < req.GetEnd() {` into `if start <= req.GetEnd() {`?

**Review by pracucci — APPROVED** [review]
The fix LGTM. I left a couple of non-blocking comments.

File: pkg/querier/queryrange/results_cache.go (hunk 2 above)
Comment (gouthamve) [reply pkg/querier/queryrange/results_cache.go:514]: Then we will always make an additional request if have an extent that ends at `req.GetEnd()`

Comment (pracucci) [reply pkg/querier/queryrange/results_cache.go:514]: Right, got it.

**Comment by bboreham** [comment]
[sorry, posted this in the wrong place yesterday]

Wouldn't it look better if you add something like this at the top and don't change the loop?

```
  // special-case situation created by step bigger than range
  if req.GetStart() == req.GetEnd() {
    // do we have a cached extent that matches this zero-length request?
    if len(extents) == 1 && extents[0].Start == req.GetStart() && extents[0].End == req.GetEnd() {
      res, err := extent.toResponse()
      cachedResponses = append(cachedResponses, res)
    } else {
      // cache didn't match this request; pass it through to querier
      requests = append(requests, req)
    }
    return requests, cachedResponses, nil
  }
```

**Comment by gouthamve** [comment]
But that would mean queries with the same start and end time will never be cached. Might not be an issue though.

**Comment by bboreham** [comment]
I'm not seeing that.  All I meant was to pull the special case out of the loop so we don't have to stare really hard at the loop to see what it means.

**Comment by bboreham** [comment]
Oh, you mean if we have a zero-length request and a non-zero-length cache entry (but still over 5 minutes) that covers it, then we won't look at that.  Fair point.

I think this helps to prove that this routine is too complicated to get right.

**Pull request merged**